utils/build_datasets.py

```diff
@@ -0,0 +1,167 @@
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-d",
+        type=str,
+        help="Path or Table Synapse Id associated with a Data Sharing Plan.",
+        required=True,
+        default=None,
+    )
+    parser.add_argument(
+        "-n",
+        type=str,
+        help="Name to apply to an updated Data Sharing Plan CSV",
+        required=False,
+        default="updated_dsp",
+    )
+    return parser.parse_args()
+
+
+def get_table(syn, source_id: str) -> pd.DataFrame:
+    """Collect a Synapse table entity and return as a Dataframe."""
+
+    query = f"SELECT * FROM {source_id}"
+    table = syn.tableQuery(query).asDataFrame().fillna("")
+
+    return table
+
+
+def filter_files_in_folder(syn, scope: str, formats: list[str]) -> list:
+    """Capture all files in provided scope and select files that match a list of formats,
+    return list of dataset items"""
+
+    all_files = []
+    walk_path = synapseutils.walk(syn, scope, ["file"])
+    for dirpath, dirname, filename in walk_path:
+        all_files = all_files + filename
+    filtered_files = [
+        file[1] for file in all_files for format in formats if file[0].endswith(format)
+    ]  # only select files of desired format
+    dataset_items = [
+        {"entityId": f, "versionNumber": syn.get(f, downloadFile=False).versionLabel}
+        for f in filtered_files
+    ]
+
+    return dataset_items
+
+
+def create_dataset_entity(syn, name: str, grant: str) -> Dataset:
+    """Create an empty Synapse Dataset using the
+    Project associated with the applicable grant number as parent.
+    Return the Dataset object."""
+
+    query = f"SELECT grantId FROM syn21918972 WHERE grantViewId='{grant}'"
+    project_id = syn.tableQuery(query).asDataFrame().iat[0, 0]
+    dataset = Dataset(name=name, parent=project_id)
+
+    return dataset
+
+
+def main():
+
+    syn = synapseclient.login()
+
+    args = get_args()
+
+    dsp, new_name = args.d, args.n
+    update_dsp_sheet = None
+    
+    if os.path.exists(dsp):
+        dsp_df = pd.read_csv(dsp, keep_default_na=False)
+        print("\nData Sharing Plan read successfully!")
+    elif "syn" in dsp:
+        dsp_df = get_table(syn, dsp)
+        print(f"Data Sharing Plan acquired from Synapse table {dsp}!")
+    else:
+        print(
+            f"❗❗❗ {dsp} is not a valid Data Sharing Plan identifier. Please check your inputs and try again."
+        )
+        exit()
+
+    if dsp_df.iat[1, 0] == "DataDSP":
+        count = 0
+        for _, row in dsp_df.iterrows():
+            grant_id = row["GrantView Key"]
+            dataset_id = row["DatasetView Key"]
+            scope_id = row["DSP Dataset Alias"]
+            dataset_name = row["DSP Dataset Name"]
+            formats = re.split(", |,", row["DSP Dataset File Formats"])
+            level = row["DSP Dataset Level"]
+            if level not in ["Metadata", "Auxiliary", "Not Applicable"]:
```

Comment:
Alternatively, you can use do something like:

```python
            if level in ["Metadata", "Auxiliary", "Not Applicable"]:
                continue

            print(f"\nProcessing Dataset {dataset_name}")
            if len(dataset_id) > 0:
                ...
```

Where Python would just move on (`continue`) to the next iteration without doing the rest of the loop's code.   This would help reduce the number of indentation levels, since now the rest of the code can be un-indented by one level.

P.S. My personal limit on indentations is 3-4 levels -- anything beyond that, I'll start thinking that my code probably needs to split further into more functions.  But that's just me!  Go with your own preferred styling :-)